core/src/main/java/kafka/log/CleanShutdownFileHandler.java

```
@@ -0,0 +1,132 @@
+        BROKER_EPOCH;
+
+        @Override
+        public String toString() {
+            return name().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    public CleanShutdownFileHandler(String dirPath) {
+        logger = new LogContext().logger(CleanShutdownFileHandler.class);
+        try {
+            this.cleanShutdownFile = new File(dirPath, CleanShutdownFileHandler.CLEAN_SHUTDOWN_FILE_NAME);
+        } catch (Exception e) {
+            logger.warn("Fail to initiate the clean shutdown file handler: " + e);
+        }
+    }
+
+    public void write(long brokerEpoch) throws Exception {
+        if (cleanShutdownFile == null) return;
+        write(brokerEpoch, currentVersion);
+    }
+
+    // visible to test.
+    void write(long brokerEpoch, int version) throws Exception {
+        FileOutputStream os = new FileOutputStream(cleanShutdownFile);
+        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(os, StandardCharsets.UTF_8));
+        try {
+            Map<String, String> payload = new HashMap<>();
+            payload.put(Fields.VERSION.toString(), Integer.toString(version));
+            payload.put(Fields.BROKER_EPOCH.toString(), Long.toString(brokerEpoch));
+            bw.write(new ObjectMapper().writeValueAsString(payload));
+            bw.flush();
+            os.getFD().sync();
+        } finally {
+            bw.close();
+            os.close();
+        }
+    }
+
+    long read() throws Exception {
+        if (cleanShutdownFile == null) return -1L;
+        BufferedReader br = Files.newBufferedReader(cleanShutdownFile.toPath(), StandardCharsets.UTF_8);
+        long brokerEpoch = -1L;
+        try {
+            String text = br.lines().collect(Collectors.joining());
+            Map<String, String> content = new ObjectMapper().readValue(text, HashMap.class);
+            int version = Integer.parseInt(content.getOrDefault(Fields.VERSION.toString(), "-1"));
+            if (version > currentVersion || version == -1) {
```

This still means that we can't downgrade if we add a new field since the old version of the software will see a higher version. In such a case, we probably only want to fail if some of the fields in the old version are missing.

Also, when we upgrade the broker to this new version for the first time, the broker will see the old cleanShutdownFile with no version. This logic will fail the in-place upgrade.

---

I guess then we don't need the version check at least now. 
Also, with an empty file, we don't have a broker epoch to use in the registration later, the method has to return -1 to indicate an unclean shutdown anyway.